src/codegen/llvm_typesystem.rs

```diff
@@ -59,6 +66,32 @@ pub fn get_llvm_float_type<'a>(context: &'a Context, size: u32, name: &str) -> F
     }
 }
 
+/// Generates a cast from the given `value` to the given `target_type` if necessary and returns the casted value. It returns
+/// the original `value` if no cast is necessary or if the provided value is not eligible to be cast (to the target type or at all).
+///
+/// This function provides no additional validation or safeguards for invalid casts, as such validation is expected to be
+/// performed at the validation stage prior to code-gen.
+/// Cast instructions for values other than IntValue, FloatValue and PointerValue will simply be ignored (and the value
+/// returned unchanged). Invalid casting instructions for the above-mentioned values will fail spectacularly instead.
+///
+/// - `llvm` the llvm utilities to use for code-generation
+/// - `index` the current Index used for type-lookups
+/// - `llvm_type_index` the type index to lookup llvm generated types
+/// - `target_type` the expected target type of the value
+/// - `value_type` the current type of the given value
+/// - `value` the value to (maybe) cast
+pub fn cast<'ctx>(
```

Comment:
I might be missing something, but can't you just change the signature of the `cast` function from `cast(llvm, index, llvm_type_index, ...)` to `cast(generator, ...)` effectively elimating the use-case of the macro?

Reply:
I have tried that. But then the generator needs to be generic (can be ExpressionCodeGenerator or StatementGenerator) and you are not easily able to access the struct fields of a generic parameter. I managed to get it working by implementing traits, but it seemed overly complicated to me when a simple wrapper macro also does the trick.

The generic implementation can still be found in [this branch](https://github.com/PLC-lang/rusty/tree/generic-cast-if-needed-experiment)

Reply:
Makes sense, LGTM 👍